### DISCUSSION TITLE
Is there  a way to use `sbt run` with all its niceness but also load the application eagerly?

### DISCUSSION THREAD

**Discussion by hilcode** [playframework/playframework#11294] [Q&A]
I really like being able to see the console output and being able to kill the app by just hitting enter but it's super annoying to have to make a request just so that the application actually starts.

**Comment by wsargent** [comment]
You can use `sbt runProd`.

**Reply by hilcode** [reply]
Unfortunately, `runProd` is nothing like `run`: it starts the app in the background (making it a pain to stop) and shows no console output.

**Comment by francisdb** [comment]
and https://github.com/spray/sbt-revolver

**Reply by hilcode** [reply]
Mmm, I don't really want to add a plugin for this and it changes how the app behaves. I have no problem killing and restarting the app, the pain is that it doesn't load eagerly.

Still, thank you for the pointer.

**Comment by francisdb** [comment]
There is a very old related ticket
https://github.com/playframework/playframework/issues/4967

I wonder what part of the play code exactly is triggering the application load after the first request

**Reply by hilcode** [reply]
Too bad that ticket just got closed without providing a good reason or an alternative. :-(

It would _seem_ simple enough to add an `--eager` flag or something like that to `sbt run`?

(Mind you, I don't even understand what the benefits are of not running the initialization code: the app is useless until that happens whether you run in "dev" or "prod" mode.)

**Reply by francisdb** [reply]
My guess is that it's done to avoid continuous compiling/restarting when not required. Could be a heavy hit on a laptop battery for example.
But I also hit this in the past where I wanted to test app startup and had to force a request every time.

**Comment by mkurz** [comment]
In dev mode, the code that triggers an application reload / (re-)compile can be found here:
https://github.com/playframework/playframework/blob/18bf4b0e4222f5f9f816178cb0e344c4adedf545/transport/server/play-server/src/main/scala/play/core/server/DevServerStart.scala#L179-L197
So what happens is that in a server backend this `applicationProvider.get` is called, which in DEV mode checks if a file was modified and eventually a re-compile happens: https://github.com/playframework/playframework/blob/18bf4b0e4222f5f9f816178cb0e344c4adedf545/transport/server/play-akka-http-server/src/main/scala/play/core/server/AkkaHttpServer.scala#L302
As you can see this happens inside the `handleRequest` method.
So to load/compile the app eagerly, this get method(or the `reload` method?) has to be called once. I am not sure if it is as easy as that, but it would be worth a try. Also, I would prefer if that would be not the default, but only when running Play with a flag (or sbt setting).

Feel free to give it a shot and provide a pull request 😉

**Reply by hilcode** [reply]
Thank you. I did see this code but I will explore a bit more. And yes, I agree that we should not just change the default behaviour. :-)

**Reply by arass** [reply]
I'm running into the same problem (on startup Play just waits for a web hit). 
Really don't want to have to hit service with a web browser since it's just backend services. Super annoying behavior when you develop.

Has this been improved? Is there an easy switch now? 
I tried searching and AIing and nothing worked.

**Reply by arass** [reply]
Yeah, but if someone did this  for my production - I would fire them. [and I don't wanna be associated with such terrible hacks]

(when I try to do runProd - it says "Bad root server path: /app/target/universal/stage/--add-opens". Googling this error wasn't helpful. so I don't wanna deal with it - I just want a "service framework" to support "services". ;) )

**Reply by codetojoy** [reply]
> Yeah, but if someone did this for my production - I would fire them. [and I don't wanna be associated with such terrible hacks]
> 

For others, this is in response to a comment that I've since deleted. It was intended for local development (the original context, re: "annoying ... when you develop") and not production.

**Comment by hilcode** [comment]
I did not look into changing anything on the Play side, I simply
created a "run-app" script that calls another script (in the
background) that repeatedly calls our health API endpoint until that
succeeds, followed by "sbt run".

So effectively this:

#!/usr/bin/bash
check-health &
sbt run

(Obviously, "check-health" sleeps between calls and there is a maximum
number of calls).

This makes it act like "sbt run" (it auto reloads and you can just hit
a key to stop the app, et cetera) but because of the repeated calls to
the health endpoint, it immediately forces the complete load of the
app. I still don't understand why this is not the default behaviour
but this workaround works well enough for me.

**Reply by arass** [reply]
Oh, yeah, that's what codejoy thought of too. But this is very hacky and while ok for a dev box to have a terminal open with a ping every few seconds - this is a horrid way to deploy in production. I'm looking for a production deploy option.

Now even the compile doesn't happen till you ping the instance. So whole thing is in a schrodinger's cat stage till someone uses it. Not very adult.

**Comment by hilcode** [comment]
I am not sure I'm following you here. Surely, you're not compiling on
a PRD instance?

For PRD, you run `sbt dist` (on your build server) and you package up
the result (RPM, tarball, zip, or whatever you like). _That_ is what
you deploy to PRD. You should not have SBT, Git, your source code, et
cetera on your PRD instance.

Or am I completely misunderstanding you? :-)

**Reply by arass** [reply]
Yes, I dist. But going back to what you said: I don't want sbt in prod? Why not? 
I just do: sbt run
But how do I ensure that something else will do evictions in the exact same way? 
sbt gives me a consistent environment across everywhere.

**Reply by HilcoWijbenga** [reply]
Generally speaking, you don't want anything other than the essentials on your PRD instances. They should be as mean and lean as possible.

You _definitely_ do not want development tools (or even a development JDK) on your PRD instance. That is a security risk.

Having said that, how much you care about security is completely up to you and depends on your situation. But we are deep in CYA territory... :-)

This is what we do (effectively):

`java $JAVA_OPTS --class-path=... play.core.server.ProdServerStart`

I believe, this exhibits the behaviour you are looking for.

**Reply by arass** [reply]
My issue with startups was that Guice was blowing up randomly cause eviction wasn't working the same way in Docker as it was locally. So I went with decision to run as SBT. But then it sits there waiting for a kick before starting up. 

Hence the desire to just have the service start fully. Seems like a bare minimal functionality (to just launch stuff)

PS: Java world has LONG abandoned not having a javac on servers. By the time someone takes your server over - javac is the least of your problems. ;)

**Comment by HilcoWijbenga** [comment]
I see, that is deeply unfortunate. This is where "it depends on your situation" comes in, obviously.

Have you compared the results of running `sbt run` and `sbt dist`? After you run `sbt dist` (or `sbt run`), the evictions are a thing of the past; they are a build time concept. What _might_ (also?) be relevant, is the order of the JARs on the classpath. This too you can control with the above `java` command.

Regarding your "P.S.", sure, I agree that the choice of dev-JDK vs non-dev-JDK is fairly minor in the grand scheme of things. That's why we take numerous steps to secure our instances. Install as few packages as possible and create as many hurdles as possible for anyone trying to do something nefarious.

**Reply by arass** [reply]
Hilco, thanks, but the simple question (which may have an answer of NO) is:
can I **sbt run** and have it not just hang there like a dead toad? (without hacking URL requests)
If one can't - I will remove Play and never go back.

**Reply by HilcoWijbenga** [reply]
To the best of my knowledge, the answer (right now) is indeed "no".

**Reply by arass** [reply]
Thanks. I'll give it a little bit in a hope someone knows a way.
